src/main/java/org/folio/rest/impl/ItemBatchSyncAPI.java

```diff
@@ -41,15 +51,27 @@ public void postItemStorageBatchSynchronous(ItemsPost entity, Map<String, String
       futures.add(setHrid(item, hridManager));
     }
 
-    CompositeFuture.all(futures).setHandler(ar -> {
-      if (ar.succeeded()) {
+    CompositeFuture.all(futures)
+      .compose(result -> effectiveCallNumberService.populateEffectiveCallNumberComponents(items))
+      .map(result -> {
         StorageHelper.postSync(ItemStorageAPI.ITEM_TABLE, entity.getItems(),
-            okapiHeaders, asyncResultHandler, vertxContext,
-            PostItemStorageBatchSynchronousResponse::respond201);
-      } else {
-        asyncResultHandler.handle(
-            Future.succeededFuture(respond500WithTextPlain(ar.cause().getMessage())));
-      }
+          okapiHeaders, asyncResultHandler, vertxContext,
+          PostItemStorageBatchSynchronousResponse::respond201);
+        return result;
+      }).otherwise(error -> {
+        log.warn("Error occurred during batch item create", error);
+
+        if (error instanceof ValidationException) {
+          final ValidationException validationError = (ValidationException) error;
+
+          asyncResultHandler.handle(Future.succeededFuture(
+            respond422WithApplicationJson(validationError.getErrors())));
+        } else {
+          asyncResultHandler.handle(Future.succeededFuture(
+            respond500WithTextPlain(error.getMessage())));
+        }
+
+        return null;
```

Comment:
Is this likely to be a common behaviour? 

I think it might be. Even if it isn't, I think it could be worth extracting it to a method that describes the intent.

Reply:
Yes it is, moved to a separate method.